### PR TITLE
<xtree>: Fixes _Tree move constructor

### LINKED ISSUE #1037: <xtree>:  _Tree move constructor incorrectly swaps
**Describe the bug**
The move constructor for a `std::set` with a lambda comparator type fails to compile when it attempts to swap the non-move-assignable comparator.

**Command-line test case**
```

d:\Temp2>type repro.cpp
#include <set>
#include <type_traits>

static const auto cmp = [](auto const& a, auto const& b) { return a < b; };

auto f()
{
    std::set<int, std::remove_const_t<decltype(cmp)>> container{cmp};
    return container; // dead at this return.
}

int main()
{
}

d:\Temp2>cl /EHsc repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\utility(107): error C2280: '<lambda_076e960ca186bb46139689bab3007458> &<lambda_076e960ca186bb46139689bab3007458>::operator =(const <lambda_076e960ca186bb46139689bab3007458> &)': attempting to reference a deleted function
repro.cpp(4): note: see declaration of '<lambda_076e960ca186bb46139689bab3007458>::operator ='
repro.cpp(4): note: '<lambda_076e960ca186bb46139689bab3007458> &<lambda_076e960ca186bb46139689bab3007458>::operator =(const <lambda_076e960ca186bb46139689bab3007458> &)': function was explicitly deleted
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\utility(114): note: see reference to function template instantiation 'void std::swap<_Ty,0>(_Ty &,_Ty &) noexcept(false)' being compiled
        with
        [
            _Ty=<lambda_076e960ca186bb46139689bab3007458>
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\xtree(994): note: see reference to function template instantiation 'void std::_Swap_adl<<lambda_076e960ca186bb46139689bab3007458>>(_Ty &,_Ty &) noexcept(false)' being compiled
        with
        [
            _Ty=<lambda_076e960ca186bb46139689bab3007458>
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\xtree(990): note: while compiling class template member function 'void std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Swap_val(std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>> &)'
        with
        [
            _Kty=int,
            _Pr=<lambda_076e960ca186bb46139689bab3007458>,
            _Alloc=std::allocator<int>
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\xtree(911): note: see reference to function template instantiation 'void std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Swap_val(std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>> &)' being compiled
        with
        [
            _Kty=int,
            _Pr=<lambda_076e960ca186bb46139689bab3007458>,
            _Alloc=std::allocator<int>
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\set(54): note: see reference to class template instantiation 'std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>' being compiled
        with
        [
            _Kty=int,
            _Pr=<lambda_076e960ca186bb46139689bab3007458>,
            _Alloc=std::allocator<int>
        ]
repro.cpp(8): note: see reference to class template instantiation 'std::set<int,<lambda_076e960ca186bb46139689bab3007458>,std::allocator<int>>' being compiled
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\utility(108): error C2280: '<lambda_076e960ca186bb46139689bab3007458> &<lambda_076e960ca186bb46139689bab3007458>::operator =(const <lambda_076e960ca186bb46139689bab3007458> &)': attempting to reference a deleted function
repro.cpp(4): note: see declaration of '<lambda_076e960ca186bb46139689bab3007458>::operator ='
repro.cpp(4): note: '<lambda_076e960ca186bb46139689bab3007458> &<lambda_076e960ca186bb46139689bab3007458>::operator =(const <lambda_076e960ca186bb46139689bab3007458> &)': function was explicitly deleted
```

**Expected behavior**
Should compile

**STL version**
```
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1
```

**Additional context**
Also tracked by DevCom-657074 and VSO-951679 / [AB#951679](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/951679)

### PR DESCRIPTION
**Pull request by Arzaghi**
Fixes #1037

### CODE REVIEW

**Comment by miscco** [comment]
From what I read in the bug description we should also fiy the move assignment.

**Comment by StephanTLavavej** [comment]
Thanks! I've pushed a batch of changes:

* Fix typo: "constuctible" => "constructible"
* Change `std::` to `_STD` in product code.
  + In addition to being conventional, `_STD` expands to `::std::` which avoids any possible ambiguity.
  + Superseded by later changes.
* In `_Move_assign` overloads, restore comparator copy assignment.
  + It's okay for container assignment to require comparator assignment.
  + Destroy-and-reconstruct is exception-unsafe - if reconstructing the comparator throws an exception (e.g. because the comparator attempts to allocate memory), then the container will be left with a destroyed comparator, violating its invariants.
  + It's intentional that container move construction/assignment performs comparator copy construction/assignment, respectively. The Standard (annoyingly!) doesn't talk about this clearly, but it's necessary to preserve container invariants. The issue is that moved-from containers still need to be in a valid-but-unspecified state (i.e. preconditionless member functions can be called), so they can be `clear()`ed (which is preconditionless) followed by inserting elements. This requires their comparators to still be usable. However, a moved-from comparator may not be usable - in particular, moved-from `std::function`s may be empty (and are, in our implementation). Therefore, because comparators must always be usable by the associative containers, they can't be moved-from.
  + This was unclear from the way I wrote up GH-1037; there I said "`_Move_assign` also copy-assigns the comparator, then calls `_Swap_val` anyways." and I should have said that the former was intentional but the latter was undesirable.
* Centralize code into `_Swap_val_excluding_comp`.
  + This code was somewhat lengthy and repeated 3 times, making it worth extracting into a helper function.
* Centralize code even more by removing `_Swap_val`.
  + This swaps the comparator in a slightly different order, but that's unobservable.
* Cleanup: Add top-level `const` to `_Right_scary` (a pointer).
* Comment when we're intentionally copying comparators.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/xtree [hunk 1]
```diff
@@ -905,10 +905,10 @@ public:
     }
 
     _Tree(_Tree&& _Right)
-        : _Mypair(
-            _One_then_variadic_args_t{}, _Right.key_comp(), _One_then_variadic_args_t{}, _STD move(_Right._Getal())) {
+        : _Mypair(_One_then_variadic_args_t{}, _Right.key_comp(), // intentionally copy comparator
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/xtree]: Thanks - I commented on that issue 7 years ago in Bristol and had forgotten that it existed.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this longstanding bug (present since at least VS 2015 and possibly all the way back to VS 2010) and improving the usability of the associative containers! :smiley_cat:

**Pull request merged**